VERSIONING.md

```diff
@@ -0,0 +1,108 @@
+# Versioning
+
+## SemVer and shorter SemVer
+
+We use [semantic versioning] for all of our images. The images with fully
+specified version (e.g. `vX.Y.Z`) are immutable and will never be changed. We
+also support a shorter version of semantic versioning. E.g.
+`v<major>` and `v<major>.<minor>`.
+
+The shorter version of semantic versioning is going to be floating tags:
+
+- `v<major>.<minor>` always points to `v<major>.<minor>.X` where X is the latest
+  (largest) patch version number. E.g. assume `v1.2` points `v1.2.0` initially.
+  After `v1.2.1` is released, `v1.2` now points to `v1.2.1`.
+
+- `v<major>` always points to `v<major>.X` where X is the latest (largest) minor
+  version number. E.g. assume `v1` points `v1.2.0` initially. After `v1.3.0` is
+  released, `v1` now points to `v1.3.0`. After `v1.3.1` is released, `v1` now
+  points to `v1.3.1`.
+
+Note: We do NOT support the `latest` tag, since we cannot provide any
+compatibility guarantee for it, and the pipeline won’t produce deterministic
+results.
+
+## User-facing Surfaces
+
+All functions in the catalog comply with the [function spec]. To learn more
+about the functions concept,
+see [here](http://kpt.dev/book/02-concepts/03-functions).
+
+There are 2 user-facing surfaces for a function:
+
+- The `functionConfig`
+- The function itself
+
+### functionConfig Surface
+
+A `functionConfig` can be either a core resource (e.g. `ConfigMap`) or a custom
+resource. If the `functionConfig` is a CRD, it can be versioned independently as
+a normal CRD.
+
+### Function Surface
+
+#### What are NOT part of the function surface
+
+- The formatting of serialization for output items and results. e.g. yaml
+  indentation and order of fields in a map.
+- The order of resources in the output items.
+- The order of result items in the results.
+- The content of the unstructured messages in the results.
+
+#### What are part of the function surface
+
+- The supported fields in the ConfigMap as the `functionConfig`.
+- The supported versions of the CRD `functionConfigs`.
+- How the function behave given the input items and `functionConfig`:
+    - The reminder aspects of the output items that are not mentioned in the
+      previous section.
+    - The reminder aspects of the results that are not mentioned in the previous
+      section.
+
+### Breaking Changes
+
+We define a breaking change as: For any given input (including `input items`,
+`functionConfig` and OpenAPI), the function produces a different output
+(including output items, results) that are part of the user-facing surface.
+
+### Backwards Compatibility
+
+For versions after v1.0.0, we will:
```

Comment:
I think adding a note that describe the semver semantics for pre 1.0 versions will be useful because all of our function's versions are currently pre-1.0.